fiwaremurano/meta/io.murano/Classes/resources/Instance.yaml

```diff
@@ -0,0 +1,292 @@
+      - $.networks.customNetworks.select($this.joinNet($, $securityGroupName))
+
+      - $preparedUserData: $.prepareUserData()
+      - $preparedUserDataCloudConfig: $.prepareUserDataCloudConfig()
+      - $shellScript: $preparedUserData.data
+      - $cloudConfig: $preparedUserDataCloudConfig.data
+      - $userDataFormat: $preparedUserData.format
+      # Create MQ queue to communicate with the VM
+      - $.agent.prepare()
+      - $template:
+          resources:
+            boot_config:
+              type: OS::Heat::CloudConfig
+              properties:
+                cloud_config:
+                  $cloudConfig
+            boot_script:
+              type: OS::Heat::SoftwareConfig
+              properties:
+                group: ungrouped
+                config: $shellScript
+            user_data:
+              type: OS::Heat::MultipartMime
+              properties:
+                parts:
+                - config: {get_resource: boot_config}
+                - config: {get_resource: boot_script}
+            $.name:
+              type: 'OS::Nova::Server'
+              properties:
+                flavor: $.flavor
+                image: $.image
+                availability_zone: $.availabilityZone
+                user_data: {get_resource: user_data}
+                user_data_format: RAW
+                key_name: $.keyname
+
+          outputs:
+            format('{0}-assigned-ips', $.name):
+              description: format('Network IPs assigned to {0} instance', $.name)
+              value:
+                get_attr: [ $.name, networks ]
+            format('{0}-id', $.name):
+              description: format('ID of {0} instance', $.name)
+              value:
+                get_resource: $.name
+
+      - $.instanceTemplate: $.instanceTemplate.mergeWith($template)
+
+      # Any additional template preparation
+      - $.instanceTemplate: $.prepareStackTemplate($.instanceTemplate)
+      - $.environment.stack.updateTemplate($.instanceTemplate)
+      - $.environment.stack.push()
+      - $outputs: $.environment.stack.output()
+      - $.ipAddresses: $outputs.get(format('{0}-assigned-ips', $this.name)).values().flatten().distinct()
+      - $.openstackId: $outputs.get(format('{0}-id', $this.name))
+      - If: $._floatingIpOutputName != null
+        Then:
+          - $.floatingIpAddress: $outputs.get($._floatingIpOutputName)
+          - If: $.floatingIpAddress != null
+            Then: $.setAttr(fipAssigned, true)
+      - $.environment.instanceNotifier.trackCloudInstance($this)
+
+  detectPrimaryNetwork:
+    Body:
+      - $._primaryNetwork: null
+      - If: $.networks.primaryNetwork != null
+        Then:
+          - $._primaryNetwork: $.networks.primaryNetwork
+        Else:
+          - If: $.networks.useEnvironmentNetwork and $.environment.defaultNetworks.environment!=null
+            Then:
+              - $._primaryNetwork: $.environment.defaultNetworks.environment
```

Comment:
IMHO, I would not use defaultNetworks.environment to reflect the network that I am using by default. I mean you are using twice the name environment and It could produce some confusion when people try to understand what are you doing here.